modules/src/ics03_connection/connection.rs

```diff
@@ -16,43 +17,31 @@ pub struct ConnectionEnd {
     versions: Vec<String>,
 }
 
-impl ConnectionEnd {
-    pub fn new(
-        client_id: ClientId,
-        counterparty: Counterparty,
-        versions: Vec<String>,
-    ) -> Result<Self, Error> {
-        Ok(Self {
-            state: State::Uninitialized,
-            client_id,
-            counterparty,
-            versions: validate_versions(versions).map_err(|e| Kind::InvalidVersion.context(e))?,
-        })
-    }
+impl TryFrom<ProtoConnectionEnd> for ConnectionEnd {
+    type Error = anomaly::Error<Kind>;
 
-    pub fn set_state(&mut self, new_state: State) {
-        self.state = new_state;
-    }
-
-    pub fn from_proto_connection_end(pc: ProtoConnectionEnd) -> Result<Self, Error> {
-        if pc.id == "" {
+    fn try_from(value: ProtoConnectionEnd) -> Result<Self, Self::Error> {
+        // Todo: Is validation complete here? (Code was moved from `from_proto_connection_end`.)
+        if value.id == "" {
```

Comment:
Following the recent changes in the code, this `if` does not make sense anymore.

I think we can now delete this `if` statement and remove the todo. The caller of `try_from` is `relayer/relay/src/query.rs:query()` and this should figure out that the response `value` is `null` and consequently return the error `ConnectionNotFound` (without even reaching the call to `try_from`). Relevant code:

https://github.com/informalsystems/ibc-rs/blob/85a94394d7cba7beac6b2b305b3c160ef6ed3ecb/relayer/relay/src/query.rs#L74-L77

Reply:
I made a commit to capture concretely my suggestion from the comment above. Full commit details [here](https://github.com/informalsystems/ibc-rs/pull/152/commits/980d0391e9b2c33f67834a4074c6627046aeb3bd).

Relevant code:

https://github.com/informalsystems/ibc-rs/blob/980d0391e9b2c33f67834a4074c6627046aeb3bd/relayer/relay/src/query.rs#L64-L67

@greg-szabo if this makes sense with what you have in mind, then we can also delete the `if` above.

Reply:
Nice, thanks! I submitted a followup that fixes the Clippy warning.

I'm not sure how much we want to manually check different results of the response before we return, but this case seems straightforward.

There's a TODO in the type validation which I think should go into it's own issue. There's a lot to improve on that piece of code.

Reply:
Awesome!
That works, we can leave the TODOs for the moment (also in `Counterparty`), and fix validation, perhaps add also some unit tests in a separate issue.